File: java/com/joshmanisdabomb/lcc/block/ClassicChestBlock.java
package com.joshmanisdabomb.lcc.block;

import com.joshmanisdabomb.lcc.misc.AdaptedFromSource;
import com.joshmanisdabomb.lcc.tileentity.ClassicChestTileEntity;
import net.minecraft.block.*;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.passive.CatEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.entity.player.ServerPlayerEntity;
import net.minecraft.inventory.IInventory;
import net.minecraft.inventory.InventoryHelper;
import net.minecraft.inventory.container.INamedContainerProvider;
import net.minecraft.item.BlockItemUseContext;
import net.minecraft.item.ItemStack;
import net.minecraft.state.DirectionProperty;
import net.minecraft.state.EnumProperty;
import net.minecraft.state.StateContainer;
import net.minecraft.state.properties.BlockStateProperties;
import net.minecraft.state.properties.ChestType;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.Direction;
import net.minecraft.util.Hand;
import net.minecraft.util.Mirror;
import net.minecraft.util.Rotation;
import net.minecraft.util.math.AxisAlignedBB;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.BlockRayTraceResult;
import net.minecraft.world.IBlockReader;
import net.minecraft.world.IWorld;
import net.minecraft.world.World;
import net.minecraftforge.fml.network.NetworkHooks;

import javax.annotation.Nullable;
import java.util.List;

public class ClassicChestBlock extends ContainerBlock implements LCCBlockHelper {

    public static final DirectionProperty FACING = HorizontalBlock.HORIZONTAL_FACING;
    public static final EnumProperty<ChestType> TYPE = BlockStateProperties.CHEST_TYPE;

    public ClassicChestBlock(Properties properties) {
        super(properties);
        this.setDefaultState(this.stateContainer.getBaseState().with(FACING, Direction.NORTH).with(TYPE, ChestType.SINGLE));
    }

    @Override
    protected void fillStateContainer(StateContainer.Builder<Block, BlockState> builder) {
        builder.add(FACING, TYPE);
    }

    @Nullable
    @Override
    public TileEntity createNewTileEntity(IBlockReader worldIn) {
        return new ClassicChestTileEntity();
    }

    @Override
    public boolean hasTileEntity(BlockState state) {
        return true;
    }

    @Override
    public boolean onBlockActivated(BlockState state, World world, BlockPos pos, PlayerEntity player, Hand handIn, BlockRayTraceResult hit) {
        if (!world.isRemote) {
            TileEntity tileEntity = world.getTileEntity(pos);
            if (tileEntity instanceof INamedContainerProvider && !this.isBlocked(world, pos)) {
                NetworkHooks.openGui((ServerPlayerEntity) player, (INamedContainerProvider) tileEntity, buf -> {
                    buf.writeBlockPos(tileEntity.getPos());
                });
            } else {
                throw new IllegalStateException("Named container provider missing.");
            }
            return true;
        }
        return true;
    }

    @Override
    @AdaptedFromSource
    public BlockState updatePostPlacement(BlockState stateIn, Direction facing, BlockState facingState, IWorld worldIn, BlockPos currentPos, BlockPos facingPos) {
        if (facingState.getBlock() == this && facing.getAxis().isHorizontal()) {
            ChestType chesttype = facingState.get(TYPE);
            if (stateIn.get(TYPE) == ChestType.SINGLE && chesttype != ChestType.SINGLE && stateIn.get(FACING) == facingState.get(FACING) && this.getDirectionToAttached(facingState) == facing.getOpposite()) {
                return stateIn.with(TYPE, chesttype.opposite());
            }
        } else if (this.getDirectionToAttached(stateIn) == facing) {
            return stateIn.with(TYPE, ChestType.SINGLE);
        }

        return super.updatePostPlacement(stateIn, facing, facingState, worldIn, currentPos, facingPos);
    }

    @Override
    @AdaptedFromSource
    public BlockState getStateForPlacement(BlockItemUseContext context) {
        ChestType chesttype = ChestType.SINGLE;
        Direction direction = context.getPlacementHorizontalFacing().getOpposite();
        boolean flag = context.isPlacerSneaking();
        Direction direction1 = context.getFace();
        if (direction1.getAxis().isHorizontal() && flag) {
            Direction direction2 = this.getDirectionToAttach(context, direction1.getOpposite());
            if (direction2 != null && direction2.getAxis() != direction1.getAxis()) {
                direction = direction2;
                chesttype = direction2.rotateYCCW() == direction1.getOpposite() ? ChestType.RIGHT : ChestType.LEFT;
            }
        }

        if (chesttype == ChestType.SINGLE && !flag) {
            if (direction == this.getDirectionToAttach(context, direction.rotateY())) {
                chesttype = ChestType.LEFT;
            } else if (direction == this.getDirectionToAttach(context, direction.rotateYCCW())) {
                chesttype = ChestType.RIGHT;
            }
        }

        return this.getDefaultState().with(FACING, direction).with(TYPE, chesttype);
    }

    public Direction getDirectionToAttached(BlockState state) {
        Direction direction = state.get(FACING);
        return state.get(TYPE) == ChestType.LEFT ? direction.rotateY() : direction.rotateYCCW();
    }

    public Direction getDirectionToAttach(BlockItemUseContext context, Direction side) {
        BlockState blockstate = context.getWorld().getBlockState(context.getPos().offset(side));
        return blockstate.getBlock() == this && blockstate.get(TYPE) == ChestType.SINGLE ? blockstate.get(FACING) : null;
    }

    @Override
    @AdaptedFromSource
    public void onBlockPlacedBy(World worldIn, BlockPos pos, BlockState state, LivingEntity placer, ItemStack stack) {
        if (stack.hasDisplayName()) {
            TileEntity tileentity = worldIn.getTileEntity(pos);
            if (tileentity instanceof ClassicChestTileEntity) {
                ((ClassicChestTileEntity)tileentity).setCustomName(stack.getDisplayName());
            }
        }

    }

    @Override
    @AdaptedFromSource
    public void onReplaced(BlockState state, World worldIn, BlockPos pos, BlockState newState, boolean isMoving) {
        if (state.getBlock() != newState.getBlock()) {
            TileEntity tileentity = worldIn.getTileEntity(pos);
            if (tileentity instanceof IInventory) {
                InventoryHelper.dropInventoryItems(worldIn, pos, (IInventory)tileentity);
                worldIn.updateComparatorOutputLevel(pos, this);
            }
            super.onReplaced(state, worldIn, pos, newState, isMoving);
        }
    }

    public BlockRenderType getRenderType(BlockState p_149645_1_) {
        return BlockRenderType.MODEL;
    }

    public boolean isBlocked(IWorld world, BlockPos pos) {
        if (world.getBlockState(pos.up()).isNormalCube(world, pos.up())) return true;
        List<CatEntity> list = world.getEntitiesWithinAABB(CatEntity.class, new AxisAlignedBB((double)pos.getX(), (double)(pos.getY() + 1), (double)pos.getZ(), (double)(pos.getX() + 1), (double)(pos.getY() + 2), (double)(pos.getZ() + 1)));
        if (!list.isEmpty()) {
            for(CatEntity entity : list) {
                if (entity.isSitting()) return true;
            }
        }
        return false;
    }

    public int getComparatorInputOverride(BlockState blockState, World worldIn, BlockPos pos) {
        return 0;//Container.calcRedstoneFromInventory(this.get(blockState, worldIn, pos, false, INVENTORY));
    }

    public BlockState rotate(BlockState state, Rotation rot) {
        return state.with(FACING, rot.rotate(state.get(FACING)));
    }

    public BlockState mirror(BlockState state, Mirror mirrorIn) {
        return state.rotate(mirrorIn.toRotation(state.get(FACING)));
    }

}

File: lcc-content/src/main/kotlin/com/joshmanisdabomb/lcc/block/ClassicChestBlock.kt
package com.joshmanisdabomb.lcc.block

import com.joshmanisdabomb.lcc.block.entity.ClassicChestBlockEntity
import net.minecraft.block.Block
import net.minecraft.block.BlockRenderType
import net.minecraft.block.BlockState
import net.minecraft.block.BlockWithEntity
import net.minecraft.block.enums.ChestType
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.mob.PiglinBrain
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemPlacementContext
import net.minecraft.item.ItemStack
import net.minecraft.screen.ScreenHandler
import net.minecraft.state.StateManager
import net.minecraft.state.property.Properties.CHEST_TYPE
import net.minecraft.state.property.Properties.HORIZONTAL_FACING
import net.minecraft.util.*
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.world.World
import net.minecraft.world.WorldAccess

class ClassicChestBlock(settings: Settings) : BlockWithEntity(settings) {

    init {
        defaultState = stateManager.defaultState.with(HORIZONTAL_FACING, Direction.NORTH).with(CHEST_TYPE, ChestType.SINGLE)
    }

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) = builder.add(HORIZONTAL_FACING, CHEST_TYPE).let {}

    override fun createBlockEntity(pos: BlockPos, state: BlockState) = ClassicChestBlockEntity(pos, state)

    override fun onUse(state: BlockState, world: World, pos: BlockPos, player: PlayerEntity, hand: Hand, hit: BlockHitResult): ActionResult {
        return if (world.isClient) {
            ActionResult.SUCCESS
        } else {
            player.openHandledScreen(state.createScreenHandlerFactory(world, pos) ?: return ActionResult.SUCCESS)
            PiglinBrain.onGuardedBlockInteracted(player, true)
            ActionResult.CONSUME
        }
    }

    fun getDirectionToAttached(state: BlockState): Direction {
        val direction = state[HORIZONTAL_FACING]
        return if (state[CHEST_TYPE] === ChestType.LEFT) direction.rotateYClockwise() else direction.rotateYCounterclockwise()
    }

    fun getDirectionToAttach(context: ItemPlacementContext, side: Direction): Direction? {
        val blockstate = context.world.getBlockState(context.blockPos.offset(side))
        return if (blockstate.block === this && blockstate[CHEST_TYPE] === ChestType.SINGLE) blockstate[HORIZONTAL_FACING] else null
    }

    override fun getStateForNeighborUpdate(state: BlockState, direction: Direction, stateFrom: BlockState, world: WorldAccess, pos: BlockPos, posFrom: BlockPos): BlockState {
        if (stateFrom.block === this && direction.axis.isHorizontal) {
            val chesttype = stateFrom[CHEST_TYPE]
            if (state[CHEST_TYPE] === ChestType.SINGLE && chesttype !== ChestType.SINGLE && state[HORIZONTAL_FACING] === stateFrom[HORIZONTAL_FACING] && this.getDirectionToAttached(stateFrom) === direction.opposite) {
                return state.with(CHEST_TYPE, chesttype.opposite)
            }
        } else if (this.getDirectionToAttached(state) === direction) {
            return state.with(CHEST_TYPE, ChestType.SINGLE)
        }

        return super.getStateForNeighborUpdate(state, direction, stateFrom, world, pos, posFrom)
    }

    override fun getPlacementState(ctx: ItemPlacementContext): BlockState? {
        var chesttype = ChestType.SINGLE
        var direction = ctx.playerFacing.opposite
        val flag = ctx.shouldCancelInteraction()
        val direction1 = ctx.side
        if (direction1.axis.isHorizontal && flag) {
            val direction2 = getDirectionToAttach(ctx, direction1.opposite)
            if (direction2 != null && direction2.axis !== direction1.axis) {
                direction = direction2
                chesttype = if (direction2.rotateYCounterclockwise() === direction1.opposite) ChestType.RIGHT else ChestType.LEFT
            }
        }

        if (chesttype === ChestType.SINGLE && !flag) {
            if (direction === getDirectionToAttach(ctx, direction.rotateYClockwise())) {
                chesttype = ChestType.LEFT
            } else if (direction === getDirectionToAttach(ctx, direction.rotateYCounterclockwise())) {
                chesttype = ChestType.RIGHT
            }
        }

        return defaultState.with(HORIZONTAL_FACING, direction).with(CHEST_TYPE, chesttype)
    }

    override fun onPlaced(world: World, pos: BlockPos, state: BlockState, placer: LivingEntity?, itemStack: ItemStack) {
        if (itemStack.hasCustomName()) (world.getBlockEntity(pos) as? ClassicChestBlockEntity)?.customName = itemStack.name
    }

    override fun onStateReplaced(state: BlockState, world: World, pos: BlockPos, newState: BlockState, moved: Boolean) {
        if (!state.isOf(newState.block)) {
            ItemScatterer.spawn(world, pos, (world.getBlockEntity(pos) as? ClassicChestBlockEntity)?.inventory ?: return super.onStateReplaced(state, world, pos, newState, moved))
            world.updateComparators(pos, this)
            if (state[CHEST_TYPE] != ChestType.SINGLE) {
                val other = getDirectionToAttached(state)
                if (world.getBlockState(pos.offset(other)).isOf(this)) world.updateComparators(pos.offset(other), this)
            }
        }
        super.onStateReplaced(state, world, pos, newState, moved)
    }

    override fun hasComparatorOutput(state: BlockState) = true

    override fun getComparatorOutput(state: BlockState, world: World, pos: BlockPos): Int {
        return ScreenHandler.calculateComparatorOutput(world.getBlockEntity(pos))
    }

    override fun getRenderType(state: BlockState) = BlockRenderType.MODEL

    override fun rotate(state: BlockState, rot: BlockRotation) = state.with(HORIZONTAL_FACING, rot.rotate(state[HORIZONTAL_FACING]))

    override fun mirror(state: BlockState, mirror: BlockMirror) = state.rotate(mirror.getRotation(state[HORIZONTAL_FACING]))

}